pkg/controller/service/service_controller.go

```diff
@@ -719,47 +608,114 @@ func (s *ServiceController) syncService(key string) error {
 
 	// service holds the latest service info from apiserver
 	service, err := s.serviceLister.Services(namespace).Get(name)
-	switch {
-	case errors.IsNotFound(err):
-		// service absence in store means watcher caught the deletion, ensure LB info is cleaned
-		glog.Infof("Service has been deleted %v. Attempting to cleanup load balancer resources", key)
-		err = s.processServiceDeletion(key)
-	case err != nil:
+	if errors.IsNotFound(err) {
+		glog.V(2).Infof("Service %q has been deleted", key)
+		return nil
+	}
+
+	if err != nil {
 		glog.Infof("Unable to retrieve service %v from store: %v", key, err)
-	default:
-		cachedService = s.cache.getOrCreate(key)
-		err = s.processServiceUpdate(cachedService, service, key)
+		return err
 	}
 
-	return err
+	if isDeletionCandidate(service) {
+		err := s.processServiceDeletion(service, key)
+		if err != nil {
+			return err
+		}
+
+		service, err = s.removeFinalizer(service)
+		if err != nil {
+			return err
+		}
+
+		// The service and all external resources, if applicable, have been removed.
+		return nil
+	}
+
+	if service.DeletionTimestamp != nil {
+		// This service has been successfully deleted so there is nothing else to do.
```

Comment:
Now that I think about it we actually don't need this check since the service will only be added to the work queue if it's marked for deletion **and** has the lb finalizer. I'll remove it.